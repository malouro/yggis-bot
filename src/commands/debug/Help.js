import Command from '../../classes/Command';

export default class Help extends Command {
	constructor({ t } = { t: () => null }) {
		super({
			name: 'Help',
			category: 'debug',
			aliases: [],
			description:
				'Help menus, support and information. Find out what commands exist and how to use them.',
			usage: {
				args: [
					{
						name: 'command',
						description: 'Get help on how to use a specific command.',
						chainable: false,
					},
					{
						name: 'commandCategory',
						description:
							'Learn more about commands available under a given category.',
						chainable: false,
					},
				],
			},
		});

		this.t = t;
		this.translationKey = 'help';

		this.name = this.t('COMMANDS', this.translationKey, 'name') || this.name;
		this.description =
			this.t('COMMANDS', this.translationKey, 'description') ||
			this.description;

		this.messageOutput = '';
	}

	getCommandArgs(command) {
		const { usage, name: commandName, translationKey } = command;
		const usageTranslations =
			this.t('COMMANDS', translationKey, 'usage') ||
			this.t('COMMANDS', commandName, 'usage');
		const chainableArgs = [];
		const nonChainableArgs = [];

		usage.args.forEach(arg => {
			if (arg.chainable) {
				chainableArgs.push(
					usageTranslations && usageTranslations[arg.name]
						? usageTranslations[arg.name].name
						: arg.name
				);
			} else {
				nonChainableArgs.push(
					usageTranslations && usageTranslations[arg.name]
						? usageTranslations[arg.name].name
						: arg.name
				);
			}
		});

		return {
			chainableArgs,
			nonChainableArgs,
		};
	}

	getCommandArgsUsage(chainableArgs, nonChainableArgs) {
		let argsUsageOutput = '';

		if (nonChainableArgs.length > 0) {
			argsUsageOutput = argsUsageOutput.concat(
				'`<',
				nonChainableArgs.join('|'),
				'>`'
			);
		}

		if (chainableArgs.length > 0) {
			argsUsageOutput = argsUsageOutput.concat(
				nonChainableArgs.length > 0 ? ' ' : '',
				'`(',
				chainableArgs.join(', '),
				')`'
			);
		}

		return argsUsageOutput;
	}

	getDetailedArgsDescription(args, commandName) {
		let argsDetailsOutput = '';

		const translations = this.t(
			'COMMANDS',
			commandName.toLocaleLowerCase(),
			'usage'
		);
		const missingDescription = this.t(
			'COMMON',
			'commands',
			'missingArgumentDescription'
		);

		args.forEach((arg, index) => {
			const description =
				(translations &&
					translations[arg.name] &&
					translations[arg.name].description) ||
				arg.description ||
				missingDescription;
			const argName =
				(translations &&
					translations[arg.name] &&
					translations[arg.name].name) ||
				arg.name;
			const argUsage =
				(translations &&
					translations[arg.name] &&
					translations[arg.name].usage) ||
				arg.usage;

			argsDetailsOutput = argsDetailsOutput.concat(
				index === 0 ? '\n' : '\n\n',
				`\`${argName}\``,
				'\n',
				description,
				arg.usage ? `${'\n'.repeat(2)}\`${argUsage}\`` : ''
			);
		});

		return argsDetailsOutput;
	}

	buildMainHelpMenu({ commandPrefix, name: botName }, categories) {
		const translate = key => this.t('COMMANDS', this.translationKey, key);
		const { chainableArgs, nonChainableArgs } = this.getCommandArgs(this);

		const mainMenuHeader =
			translate('mainMenuHeader')(botName) || `**__${botName} Help Menu__**`;
		const helpCommandUsage = String.prototype.concat(
			`\`${commandPrefix}${this.name.toLocaleLowerCase()}\``,
			' ',
			`${this.getCommandArgsUsage(chainableArgs, nonChainableArgs)}`
		);
		const commandsHeader =
			translate('mainMenuCommandsHeader') || '**Commands**';
		const categoryHeader = category =>
			translate('mainMenuCategoryHeader')(category.name) ||
			`> ${category.name} commands\n`;
		const commandList = categories
			.map(category =>
				categoryHeader(category).concat(
					category.commands
						.map(command => `\`${commandPrefix}${command}\``)
						.join(', ')
				)
			)
			.join('\n\n');

		return [
			mainMenuHeader,
			'',
			helpCommandUsage,
			'',
			commandsHeader,
			'',
			commandList,
		].join('\n');
	}

	/**
	 * This will build the text to display for any given Command's help menu
	 *
	 * The information here is generated by the Commands predefined `description`, `usage`, and `args` usage
	 */
	buildCommandHelpMenu({ commandPrefix }, command, invalidUsage = false) {
		const { usage, name, translationKey } = command;
		const { chainableArgs, nonChainableArgs } = this.getCommandArgs(command);

		const translate = key => this.t('COMMANDS', name.toLocaleLowerCase(), key);

		/* Get all appropriate text to display */
		const description = translate('description') || command.description;
		const title =
			(typeof this.t('COMMANDS', translationKey, 'commandHeader') ===
				'function' &&
				this.t('COMMANDS', translationKey, 'commandHeader')(name)) ||
			`**${name} Command**`;
		const commandUsage = String.prototype.concat(
			`\`${commandPrefix}${name.toLocaleLowerCase()}\``,
			' ',
			`${this.getCommandArgsUsage(chainableArgs, nonChainableArgs)}`
		);
		const invalidUsageText =
			(typeof this.t('COMMANDS', translationKey, 'invalidCommandUsage') ===
				'function' &&
				this.t(
					'COMMANDS',
					translationKey,
					'invalidCommandUsage'
				)(commandPrefix, command.name.toLocaleLowerCase())) ||
			`Invalid \`${commandPrefix}${command.name.toLocaleLowerCase()}\` usage.\n`;
		const missingDescriptionText =
			this.t('COMMON', 'commands', 'missingArgumentDescription') ||
			'(Missing description.)';
		const descriptionText = `${title}\n\n${description ||
			missingDescriptionText}\n`;

		return [
			invalidUsage ? invalidUsageText : descriptionText,
			'',
			commandUsage,
			'',
			this.getDetailedArgsDescription(usage.args, translationKey || name),
		].join('\n');
	}

	buildCategoryHelpMenu({ commandPrefix }, category) {
		const { name: categoryName, description, commands } = category;
		const translations = this.t(
			'COMMANDS',
			'categories',
			categoryName.toLocaleLowerCase()
		);
		const name = translations ? translations.name : categoryName;
		const title =
			(typeof this.t('COMMANDS', this.translationKey, 'categoryHeader') ===
				'function' &&
				this.t('COMMANDS', this.translationKey, 'categoryHeader')(name)) ||
			`**${name} Commands**`;

		const commandList = commands
			.map(command => `\`${commandPrefix}${command}\``)
			.join(', ');

		const descriptionText =
			translations && translations.description
				? `\n${translations.description}\n`
				: `\n${description}\n`;

		// prettier-ignore
		return [
			title,
			description ? descriptionText : '',
			commandList
		].join('\n');
	}

	/* Determine from the arguments what to actually run */
	preAction({ args, bot }) {
		if (args.length === 1) {
			// !help (no args)
			/* eslint-disable no-underscore-dangle */
			this.messageOutput = this.buildMainHelpMenu(bot, bot._commandCategories);
		} else if (bot.commands.has(args[1])) {
			// !help <command>
			this.messageOutput = this.buildCommandHelpMenu(
				bot,
				bot.commands.get(args[1].toLocaleLowerCase())
			);
		} else if (bot._commandCategories.has(args[1])) {
			// !help <category>
			this.messageOutput = this.buildCategoryHelpMenu(
				bot,
				bot._commandCategories.get(args[1].toLocaleLowerCase())
			);
		} else {
			// !help {invalidUsage}
			this.messageOutput = this.buildCommandHelpMenu(bot, this, true);
		}
		/* eslint-enable no-underscore-dangle */
	}

	action({ message }) {
		message.channel.send(this.messageOutput);
	}
}
