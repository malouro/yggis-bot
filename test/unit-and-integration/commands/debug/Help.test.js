import mockConsole from 'jest-mock-console';

import Help from '../../../../src/commands/debug/Help';
import {
	runCommand,
	makeMockBot,
	makeMockCommand,
	MockTranslateFunc,
	MockDefaultConfig,
} from '../../../testHelpers';

describe('Help Command', () => {
	let restoreConsole = null;

	beforeAll(() => {
		restoreConsole = mockConsole();
	});

	afterAll(() => {
		restoreConsole();
	});

	const HelpCommand = new Help({ t: MockTranslateFunc() });
	const mockBot = makeMockBot({ mockCommand: Help });

	/**
	 * @description
	 * - Runs the Help command, with a given set of option overrides
	 * - Use an array for `args` to list out arguments to be used in the command execution
	 * @returns Jest Mock results to the mocked out Discord `message.channel.send()`
	 */
	const runHelpCommand = async ({
		args = ['!help'],
		bot = mockBot,
		...options
	} = {}) => {
		const {
			message: {
				channel: { send: mockedSend },
			},
		} = await runCommand(HelpCommand, {
			args,
			bot,
			...options,
		});

		return mockedSend.mock.results[0].value;
	};

	afterEach(() => {
		jest.clearAllMocks();
	});

	describe('makes a main help menu', () => {
		let mainHelpMenu = null;

		beforeAll(async () => {
			mainHelpMenu = await runHelpCommand();
		});

		test('that returns a string', () => {
			expect(typeof mainHelpMenu).toBe('string');
		});

		test('that looks like this', () => {
			expect(mainHelpMenu).toMatchSnapshot();
		});

		test('that looks like this when there are multiple commands', async () => {
			const mockCommand1 = makeMockCommand({ name: 'command1' });
			const mockCommand2 = makeMockCommand({ name: 'command2' });
			const mockBotWithCommands = makeMockBot({
				commands: [Help, mockCommand1, mockCommand2],
			});

			const {
				message: {
					channel: { send: mockedSend },
				},
			} = await runCommand(HelpCommand, {
				bot: mockBotWithCommands,
				args: ['!help'],
			});

			const helpMenuResult = mockedSend.mock.results[0].value;

			expect(helpMenuResult).toMatchSnapshot();
		});

		test('when invalid usage is given', async () => {
			const invalidUsageHelpMenu = await runHelpCommand({
				args: ['!help', 'invalid'],
			});

			expect(invalidUsageHelpMenu).toMatchSnapshot();
		});
	});

	describe("makes a command's help menu", () => {
		const makeArg = (number, chainable = false, usage = null) => ({
			name: `arg${number}`,
			description: `Arg#${number}`,
			chainable,
			usage,
		});

		const commandWithSomeArgs = makeMockCommand({
			name: 'Command',
			usage: {
				args: [makeArg(1, true), makeArg(2, false), { name: 'arg3' }],
			},
		});
		const commandWithChainArgs = makeMockCommand({
			name: 'Chain',
			usage: {
				args: [makeArg(1, true), makeArg(2, true)],
			},
		});
		const commandWithNonChainArgs = makeMockCommand({
			name: 'NonChain',
			usage: {
				args: [makeArg(1, false), makeArg(2, false)],
			},
		});
		const commandWithBoth = makeMockCommand({
			name: 'Both',
			usage: {
				args: [makeArg(1, true), makeArg(2, false)],
			},
		});
		const commandWithArgUsage = makeMockCommand({
			name: 'ArgUsage',
			usage: {
				args: [
					makeArg(1, true, 'Use this like that.'),
					makeArg(2, false, 'Use that like this.'),
				],
			},
		});

		const bot = makeMockBot({
			commands: [
				Help,
				commandWithSomeArgs,
				commandWithChainArgs,
				commandWithNonChainArgs,
				commandWithBoth,
				commandWithArgUsage,
			],
			t: MockTranslateFunc(),
		});

		const chainArgRegex = /\([\w,\s]+\)/;
		const nonChainArgRegex = /<[\w|]+>/;

		test("that displays a command's available arguments", async () => {
			const commandWithSomeArgsResult = await runHelpCommand({
				bot,
				args: ['!help', 'command'],
			});

			expect(commandWithSomeArgsResult).toEqual(
				expect.stringContaining('Command')
			);
			expect(commandWithSomeArgsResult).toEqual(
				expect.stringContaining('!command')
			);
			expect(commandWithSomeArgsResult).toEqual(
				expect.stringContaining('`<arg2|arg3>` `(arg1)`')
			);
			expect(commandWithSomeArgsResult).toEqual(
				expect.stringContaining('`arg3`\n(Missing description.)')
			);

			expect(commandWithSomeArgsResult).toMatchSnapshot();
		});

		test("that displays an argument's usage details when available", async () => {
			const argUsageResult = await runHelpCommand({
				bot,
				/* cspell: disable-next-line */
				args: ['!help', 'argusage'],
			});

			expect(argUsageResult).toMatchSnapshot();
		});

		test('with chainable arguments displayed', async () => {
			const chainableArgsResult = await runHelpCommand({
				bot,
				args: ['!help', 'chain'],
			});

			expect(chainableArgsResult).toEqual(expect.stringMatching(chainArgRegex));
			expect(chainableArgsResult).not.toEqual(
				expect.stringMatching(nonChainArgRegex)
			);
		});

		test('with non-chainable arguments displayed', async () => {
			const nonChainableArgsResult = await runHelpCommand({
				bot,
				/* cspell: disable-next-line */
				args: ['!help', 'nonchain'],
			});

			expect(nonChainableArgsResult).toEqual(
				expect.stringMatching(nonChainArgRegex)
			);
			expect(nonChainableArgsResult).not.toEqual(
				expect.stringMatching(chainArgRegex)
			);
		});

		test('with both chainable & non-chainable arguments displayed', async () => {
			const bothArgTypesResult = await runHelpCommand({
				bot,
				args: ['!help', 'both'],
			});

			expect(bothArgTypesResult).toEqual(expect.stringMatching(chainArgRegex));
			expect(bothArgTypesResult).toEqual(
				expect.stringMatching(nonChainArgRegex)
			);
		});
	});

	describe("makes a command category's help menu", () => {
		test('with or without command category configuration', async () => {
			const testCommand = makeMockCommand({
				name: 'TestCommand',
				category: 'test',
			});
			const noCategoryConfigCommand = makeMockCommand({
				name: 'NoCatConfigCommand',
				category: 'no-cat',
			});

			const commands = [testCommand, noCategoryConfigCommand];

			const bot = makeMockBot({
				commands,
				...MockDefaultConfig,
				commandPrefix: '!',
				commandCategories: {
					test: {
						name: 'Test',
						description: 'Testing command categories',
					},
				},
			});

			const testCategoryHelpMenu = await runHelpCommand({
				args: ['!help', 'test'],
				bot,
			});

			jest.clearAllMocks();

			const noCategoryConfigHelpMenu = await runHelpCommand({
				args: ['!help', 'no-cat'],
				bot,
			});

			expect(testCategoryHelpMenu).toMatchSnapshot(
				'should have a category description'
			);
			expect(noCategoryConfigHelpMenu).toMatchSnapshot(
				'should not have a category description, and the name is generated'
			);
		});
	});

	describe('has its own command help menu', () => {
		let helpCommandHelpMenu = null;

		beforeAll(async () => {
			helpCommandHelpMenu = await runHelpCommand({ args: ['!help', 'help'] });
		});

		test('that looks like this', () => {
			expect(helpCommandHelpMenu).toMatchSnapshot();
		});

		test.each(
			HelpCommand.usage.args.map(arg => [arg.name, arg.description])
		)(
			"that shows the `%s` argument's description",
			(argName, argDescription = '') =>
				expect(helpCommandHelpMenu).toEqual(
					expect.stringContaining(argDescription)
				)
		);
	});

	describe('has its own command category help menu (for the `debug` category)', () => {
		let debugCategoryHelpMenu = null;

		beforeAll(async () => {
			debugCategoryHelpMenu = await runHelpCommand({
				args: ['!help', 'debug'],
			});
		});

		test('that looks like this', () => {
			expect(debugCategoryHelpMenu).toMatchSnapshot();
		});
	});
});
