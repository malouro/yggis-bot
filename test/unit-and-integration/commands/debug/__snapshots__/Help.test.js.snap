// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help Command has its own command category help menu (for the \`debug\` category) that looks like this 1`] = `
"**Debug Commands**

Commands that assist with debugging and support. üîß

\`!help\`"
`;

exports[`Help Command has its own command help menu that looks like this 1`] = `
"**Help Command**

Help menus, support and information. Find out what commands exist and how to use them. üíÅ‚Äç‚ôÄÔ∏è


\`!help\` \`<command|commandCategory>\`


\`command\`
Get help on how to use a specific command.

\`commandCategory\`
Learn more about commands available under a given category."
`;

exports[`Help Command makes a command category's help menu with or without command category configuration: should have a category description 1`] = `
"**Test Commands**

Testing command categories

\`!testcommand\`"
`;

exports[`Help Command makes a command category's help menu with or without command category configuration: should not have a category description, and the name is generated 1`] = `
"**NoCat Commands**

\`!nocatconfigcommand\`"
`;

exports[`Help Command makes a command's help menu that displays a command's available arguments 1`] = `
"**Command Command**

(Missing description.)


\`!command\` \`<arg2|arg3>\` \`(arg1)\`


\`arg1\`
Arg#1

\`arg2\`
Arg#2

\`arg3\`
(Missing description.)"
`;

exports[`Help Command makes a command's help menu that displays an argument's usage details when available 1`] = `
"**ArgUsage Command**

(Missing description.)


\`!argusage\` \`<arg2>\` \`(arg1)\`


\`arg1\`
Arg#1

\`Use this like that.\`

\`arg2\`
Arg#2

\`Use that like this.\`"
`;

exports[`Help Command makes a main help menu that looks like this 1`] = `
"**__Yggis Help Menu__**

\`!help\` \`<command|commandCategory>\`

**Commands**

> Debug commands
\`!help\`"
`;

exports[`Help Command makes a main help menu that looks like this when there are multiple commands 1`] = `
"**__Yggis Help Menu__**

\`!help\` \`<command|commandCategory>\`

**Commands**

> Misc commands
\`!command1\`, \`!command2\`

> Debug commands
\`!help\`"
`;

exports[`Help Command makes a main help menu when invalid usage is given 1`] = `
"Invalid Invalid \`help!\` usage.


\`!help\` \`<command|commandCategory>\`


\`command\`
Get help on how to use a specific command.

\`commandCategory\`
Learn more about commands available under a given category."
`;
